File: Engine/Sources/Anzan/Eval/DataType.swift
import Foundation // trimmingCharacters for the trailing # doc comment

/// A user-declared record type: `data Person { name: String, age: Number,
/// active: Boolean }`. Like `UserFunction`, the original `source` line is
/// kept for workbook serialization — including any trailing `# doc comment`,
/// which is the type's documentation.
///
/// Construction goes through the type's CONSTRUCTOR (the type name, called
/// like a function): named fields — `Person(name: "Ada", age: 36, active:
/// true)` — or one map. There is deliberately no positional form (user
/// decision: field names at every call site).
public struct DataType: Equatable, Sendable {
    /// As declared — must start with a capital letter. Constructor calls are
    /// case-insensitive, like every function.
    public let name: String
    /// Declaration order — instances canonicalize their fields to it.
    public let fields: [DataField]
    public internal(set) var source: String

    public init(name: String, fields: [DataField], source: String = "") {
        self.name = name
        self.fields = fields
        self.source = source
    }

    /// The trailing `# …` comment of the declaration, if any — the user's
    /// own documentation, shown by man()/the reference window.
    public var documentation: String? {
        guard let hash = source.firstIndex(of: "#") else { return nil }
        let text = source[source.index(after: hash)...].trimmingCharacters(in: .whitespaces)
        return text.isEmpty ? nil : text
    }

    /// Display form: `data Person { name: String, age: Number }`.
    public var declaration: String {
        "data \(name) { "
            + fields.map { "\($0.name): \($0.type.label)" }.joined(separator: ", ")
            + " }"
    }

    /// "name, age, active" — for error messages.
    var fieldList: String {
        fields.map(\.name).joined(separator: ", ")
    }
}

/// One declared field. Names are case-sensitive (they become map-style keys);
/// duplicates are rejected at parse time.
public struct DataField: Equatable, Sendable {
    public let name: String
    public let type: DataFieldType

    public init(name: String, type: DataFieldType) {
        self.name = name
        self.type = type
    }

    /// Checks a constructor argument against the declared type (recursing into
    /// list/map element types).
    func validated(_ value: Value, in typeName: String) throws -> Value {
        try type.validate(value, field: name, in: typeName)
    }
}

/// A field's type: a built-in scalar (Boolean fields hold the engine's 1/0 but
/// render/serialize as true/false), `.record(name)` — another declared data
/// type, so records nest (`data Line { a: Point, b: Point }`) — or the composite
/// `.list(T)` (`[String]`, `[[Number]]`) and `.map(T)` (`{String: Number}`, a
/// string-keyed map of `T`). `indirect` so list/map can wrap any field type.
public indirect enum DataFieldType: Equatable, Sendable {
    case number
    case string
    case boolean
    case record(String)        // a declared data type, e.g. Point
    case list(DataFieldType)   // [T]
    case map(DataFieldType)    // {String: T} — string-keyed map of T

    /// Parses a LEAF type from a single token (scalar or a data-type name); the
    /// parser handles the composite `[…]` / `{…}` forms. Returns nil otherwise.
    public init?(parsing text: String) {
        switch text.lowercased() {
        case "number": self = .number
        case "string": self = .string
        case "boolean": self = .boolean
        default:
            guard let first = text.first, first.isUppercase else { return nil }
            self = .record(text)
        }
    }

    /// Canonical spelling — `Number` / `Point` / `[String]` / `{String: Number}`.
    public var label: String {
        switch self {
        case .number: return "Number"
        case .string: return "String"
        case .boolean: return "Boolean"
        case .record(let name): return name
        case .list(let element): return "[\(element.label)]"
        case .map(let valueType): return "{String: \(valueType.label)}"
        }
    }

    /// Validates a value against this type, recursing into list/map elements.
    /// Booleans are the engine's 1/0, but a Boolean field is strict (exactly 0/1,
    /// so `active: 7` is caught). Records are already-validated immutable
    /// instances, so a type-name check suffices (no re-validation, no cycles).
    func validate(_ value: Value, field: String, in typeName: String) throws -> Value {
        func mismatch() -> EngineError {
            .domainError(message: "'\(field)' of \(typeName) is a \(label) — got \(value.kindName)")
        }
        switch self {
        case .number:
            guard case .number = value else { throw mismatch() }
        case .string:
            guard case .string = value else { throw mismatch() }
        case .boolean:
            guard case .number(let flag) = value, flag == .zero || flag == .one else {
                throw EngineError.domainError(
                    message: "'\(field)' of \(typeName) is a Boolean — use true or false")
            }
        case .record(let expected):
            guard case .record(let record) = value,
                  record.typeName.lowercased() == expected.lowercased() else { throw mismatch() }
        case .list(let element):
            guard case .array(let items) = value else { throw mismatch() }
            return .array(try items.map { try element.validate($0, field: field, in: typeName) })
        case .map(let valueType):
            guard case .map(let entries) = value else { throw mismatch() }
            return .map(try entries.map {
                Value.MapEntry(key: $0.key,
                               value: try valueType.validate($0.value, field: field, in: typeName))
            })
        }
        return value
    }
}

File: rust/anzan/src/eval/data_type.rs
//! A user-declared record type: `data Person { name: String, age: Number,
//! active: Boolean }`. Like `UserFunction`, the original `source` line is
//! kept for workbook serialization — including any trailing `# doc comment`,
//! which is the type's documentation.
//!
//! Construction goes through the type's CONSTRUCTOR (the type name, called
//! like a function): named fields — `Person(name: "Ada", age: 36, active:
//! true)` — or one map. There is deliberately no positional form (user
//! decision: field names at every call site).
//!
//! Value validation (`DataFieldType::validate`) arrives with the `Value`
//! port; the declaration side lives here because the parser produces it.

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
pub struct DataType {
    /// As declared — must start with a capital letter. Constructor calls are
    /// case-insensitive, like every function.
    pub name: String,
    /// Declaration order — instances canonicalize their fields to it.
    pub fields: Vec<DataField>,
    pub source: String,
}

impl DataType {
    pub fn new(name: impl Into<String>, fields: Vec<DataField>, source: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            fields,
            source: source.into(),
        }
    }

    /// The trailing `# …` comment of the declaration, if any — the user's
    /// own documentation, shown by man()/the reference window.
    pub fn documentation(&self) -> Option<String> {
        let hash = self.source.find('#')?;
        let text = self.source[hash + 1..].trim();
        if text.is_empty() {
            None
        } else {
            Some(text.to_string())
        }
    }

    /// Display form: `data Person { name: String, age: Number }`.
    pub fn declaration(&self) -> String {
        let fields: Vec<String> = self
            .fields
            .iter()
            .map(|f| format!("{}: {}", f.name, f.field_type.label()))
            .collect();
        format!("data {} {{ {} }}", self.name, fields.join(", "))
    }

    /// "name, age, active" — for error messages.
    pub(crate) fn field_list(&self) -> String {
        self.fields
            .iter()
            .map(|f| f.name.as_str())
            .collect::<Vec<_>>()
            .join(", ")
    }
}

/// One declared field. Names are case-sensitive (they become map-style keys);
/// duplicates are rejected at parse time.
#[derive(Debug, Clone, PartialEq)]
pub struct DataField {
    pub name: String,
    pub field_type: DataFieldType,
}

impl DataField {
    pub fn new(name: impl Into<String>, field_type: DataFieldType) -> Self {
        Self {
            name: name.into(),
            field_type,
        }
    }
}

/// A field's type: a built-in scalar (Boolean fields hold the engine's 1/0
/// but render/serialize as true/false), `Record(name)` — another declared
/// data type, so records nest (`data Line { a: Point, b: Point }`) — or the
/// composite `List(T)` (`[String]`, `[[Number]]`) and `Map(T)`
/// (`{String: Number}`, a string-keyed map of `T`).
#[derive(Debug, Clone, PartialEq)]
pub enum DataFieldType {
    Number,
    String,
    Boolean,
    /// A declared data type, e.g. Point.
    Record(String),
    /// [T]
    List(Box<DataFieldType>),
    /// {String: T} — string-keyed map of T.
    Map(Box<DataFieldType>),
}

impl DataFieldType {
    /// Parses a LEAF type from a single token (scalar or a data-type name);
    /// the parser handles the composite `[…]` / `{…}` forms. `None` otherwise.
    pub fn parsing(text: &str) -> Option<Self> {
        match text.to_lowercase().as_str() {
            "number" => Some(Self::Number),
            "string" => Some(Self::String),
            "boolean" => Some(Self::Boolean),
            _ => {
                let first = text.chars().next()?;
                if first.is_uppercase() {
                    Some(Self::Record(text.to_string()))
                } else {
                    None
                }
            }
        }
    }

    /// Canonical spelling — `Number` / `Point` / `[String]` /
    /// `{String: Number}`.
    pub fn label(&self) -> String {
        match self {
            Self::Number => "Number".to_string(),
            Self::String => "String".to_string(),
            Self::Boolean => "Boolean".to_string(),
            Self::Record(name) => name.clone(),
            Self::List(element) => format!("[{}]", element.label()),
            Self::Map(value_type) => format!("{{String: {}}}", value_type.label()),
        }
    }

    /// Within a `namespace`, qualify a record-type reference: an unqualified
    /// `Point` mapped by `scope` (a simple lowercased name → its qualified
    /// form, accumulated from the enclosing namespaces) becomes `Bits::Point`;
    /// an already-qualified (`Other::T`) or out-of-scope global name is left
    /// alone. Recurses into list/map element types.
    pub(crate) fn qualified(&self, scope: &HashMap<String, String>) -> DataFieldType {
        match self {
            Self::Number | Self::String | Self::Boolean => self.clone(),
            Self::Record(name) => {
                if !name.contains("::") {
                    if let Some(qualified) = scope.get(&name.to_lowercase()) {
                        return Self::Record(qualified.clone());
                    }
                }
                self.clone()
            }
            Self::List(element) => Self::List(Box::new(element.qualified(scope))),
            Self::Map(value_type) => Self::Map(Box::new(value_type.qualified(scope))),
        }
    }
}

// MARK: - Value validation (the constructor's type checks)

use super::value::{MapEntry, Value};
use crate::EngineError;

impl DataFieldType {
    /// Validates a value against this type, recursing into list/map
    /// elements. Booleans are the engine's 1/0, but a Boolean field is
    /// strict (exactly 0/1, so `active: 7` is caught). Records are
    /// already-validated immutable instances, so a type-name check suffices
    /// (no re-validation, no cycles).
    pub(crate) fn validate(
        &self,
        value: &Value,
        field: &str,
        type_name: &str,
    ) -> Result<Value, EngineError> {
        let mismatch = || {
            EngineError::domain(format!(
                "'{field}' of {type_name} is a {} — got {}",
                self.label(),
                value.kind_name()
            ))
        };
        match self {
            Self::Number => {
                if !matches!(value, Value::Number(_)) {
                    return Err(mismatch());
                }
            }
            Self::String => {
                if !matches!(value, Value::String(_)) {
                    return Err(mismatch());
                }
            }
            Self::Boolean => {
                let ok = matches!(value, Value::Number(flag)
                    if flag.is_zero() || *flag == crate::BigDecimal::one());
                if !ok {
                    return Err(EngineError::domain(format!(
                        "'{field}' of {type_name} is a Boolean — use true or false"
                    )));
                }
            }
            Self::Record(expected) => {
                let ok = matches!(value, Value::Record(record)
                    if record.type_name.eq_ignore_ascii_case(expected));
                if !ok {
                    return Err(mismatch());
                }
            }
            Self::List(element) => {
                let Value::Array(items) = value else {
                    return Err(mismatch());
                };
                let validated: Result<Vec<Value>, EngineError> = items
                    .iter()
                    .map(|item| element.validate(item, field, type_name))
                    .collect();
                return Ok(Value::Array(validated?));
            }
            Self::Map(value_type) => {
                let Value::Map(entries) = value else {
                    return Err(mismatch());
                };
                let validated: Result<Vec<MapEntry>, EngineError> = entries
                    .iter()
                    .map(|e| {
                        Ok(MapEntry::new(
                            e.key.clone(),
                            value_type.validate(&e.value, field, type_name)?,
                        ))
                    })
                    .collect();
                return Ok(Value::Map(validated?));
            }
        }
        Ok(value.clone())
    }
}
